actors/builtin/miner/miner_state.go

```diff
@@ -957,6 +985,134 @@ func (st *State) checkPrecommitExpiry(store adt.Store, sectors abi.BitField) (de
 	return depositToBurn, nil
 }
 
+type AdvanceDeadlineResult struct {
+	PledgeDelta                           abi.TokenAmount
+	PowerDelta                            PowerPair
+	DetectedFaultyPower, TotalFaultyPower PowerPair
+}
+
+// AdvanceDeadline advances the deadline. It:
+// - Processes expired sectors.
+// - Handles missed proofs.
+// - Returns the changes to power & pledge, and faulty power (both declared and undeclared).
+func (st *State) AdvanceDeadline(store adt.Store, currEpoch abi.ChainEpoch) (*AdvanceDeadlineResult, error) {
+	pledgeDelta := abi.NewTokenAmount(0)
+	powerDelta := NewPowerPairZero()
+
+	detectedFaultyPower := NewPowerPairZero()
+	totalFaultyPower := NewPowerPairZero()
+
+	// Note: Use dlInfo.Last() rather than rt.CurrEpoch unless certain
+	// of the desired semantics. In the past, this method would sometimes be
+	// invoked late due to skipped blocks. This is no longer the case, but
+	// we still use dlInfo.Last().
+	dlInfo := st.DeadlineInfo(currEpoch)
+
+	// This method is invoked once *before* the first proving period starts,
+	// after the actor is first constructed; this is detected by
+	// !dlInfo.PeriodStarted().
+	if !dlInfo.PeriodStarted() {
+		return &AdvanceDeadlineResult{
+			pledgeDelta,
+			powerDelta,
+			detectedFaultyPower,
+			totalFaultyPower,
+		}, nil
+	}
+
+	// Advance to the next deadline (in case we short-circuit below).
+	st.CurrentDeadline = (st.CurrentDeadline + 1) % WPoStPeriodDeadlines
+	if st.CurrentDeadline == 0 {
+		st.ProvingPeriodStart = st.ProvingPeriodStart + WPoStProvingPeriod
+	}
+
+	deadlines, err := st.LoadDeadlines(store)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load deadlines: %w", err)
+	}
+	deadline, err := deadlines.LoadDeadline(store, dlInfo.Index)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load deadline %d: %w", dlInfo.Index, err)
+	}
+
+	// No live sectors in this deadline, nothing to do.
+	if deadline.LiveSectors == 0 {
+		return &AdvanceDeadlineResult{
+			pledgeDelta,
+			powerDelta,
+			detectedFaultyPower,
+			totalFaultyPower,
+		}, nil
+	}
+
+	quant := dlInfo.QuantSpec()
+	{
+		// Detect and penalize missing proofs.
+		faultExpiration := dlInfo.Last() + FaultMaxAge
+
+		powerDelta, detectedFaultyPower, err = deadline.ProcessDeadlineEnd(store, quant, faultExpiration)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to process end of deadline %d: %w", dlInfo.Index, err)
+		}
+		totalFaultyPower = deadline.FaultyPower
```

Comment:
I would prefer to do away with this `totalFaultyPower` var completely, and just use `deadline.FaultyPower` in the return value.

Reply:
I made a really bad suggestion here. `deadline.FaultyPower` is subsequently mutated in `PopExpiredSectors`, which causes it to forget about faulty-and-expiring power.